### PR TITLE
[W5.4][W15-A2]See Jia Min Jasmine

### PR DESCRIPTION
**Pull request by JasmineSee**
I made find case-insensitive and added an Edit function that can edit details of the person through index.

### CODE REVIEW

**Comment by zameschua** [comment]
There's a lot of irrelevant changes in this PR, which makes it very hard to read! It would be much cleaner if you could create a new branch and cherry-pick only the relevant commits.

Refer to #458 for the comments that I gave Shao Cong on this, I think it should help you!

**Comment by ShaocongDong** [comment]
Hi Jasmine, you have done a very good job.
#### Implemented Functionalities - Edit
------
* Your `edit` command is successfully added.  
  * It accepts the `index`, which is easier to find from a list.
  * It has flexible syntax to accept multiple edit changes in the same request.
  * You change (edit) the person inside the addressBook instead of `delete` and `add` again, which can be more consistent and efficient.
  * It may be better if you can remove the irrelevant old codes marked as comments for now.
  * The edit argument checking is very thorough and useful.

#### General comments  
------
* Good to have several style changes including some of the method being changed to static.
* The tests are also updated, including the shell scripts tests. The Edit Command test looks very well-rounded.
* Parser for edit part is also tested using `Junit` tests.

File: src/seedu/addressbook/Main.java [hunk 1]
```diff
@@ -22,33 +28,44 @@
...
     private TextUi ui;
     private StorageFile storage;
     private AddressBook addressBook;
 
-    /** The list of person shown to the user most recently.  */
+
+    /**
+     * The list of person shown to the user most recently.
+     */
     private List<? extends ReadOnlyPerson> lastShownList = Collections.emptyList();
```
Comment (fanyiii) [inline src/seedu/addressbook/Main.java]: Both formats are ok since it's just one line

File: src/seedu/addressbook/commands/EditCommand.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
+package seedu.addressbook.commands;
+
+import seedu.addressbook.common.Messages;
+import seedu.addressbook.data.exception.IllegalValueException;
+import seedu.addressbook.data.person.Address;
+import seedu.addressbook.data.person.Person;
+import seedu.addressbook.data.person.ReadOnlyPerson;
+import seedu.addressbook.data.person.UniquePersonList.PersonNotFoundException;
+
+import java.util.*;
```
Comment (fanyiii) [inline src/seedu/addressbook/commands/EditCommand.java]: Imported classes should always be listed explicitly.

File: src/seedu/addressbook/commands/EditCommand.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+//    public EditCommand(Set<String> keywords) {
+//        this.keywords = keywords;
+//    }
+
+    /**
+     * Returns a copy of keywords in this command.
+     */
+    //  public Set<String> getKeywords() {
+//        return new HashSet<>(keywords);
+//    }
```
Comment (fanyiii) [inline src/seedu/addressbook/commands/EditCommand.java]: You should delete these unused codes to make it cleaner

File: src/seedu/addressbook/commands/EditCommand.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    public EditCommand(int targetVisibleIndex, String arg) throws IllegalValueException {
+        super(targetVisibleIndex);
+        this.targetIndex=targetVisibleIndex;
+        if(arg.length()>=2) {
+            this.arg = arg.substring(2);
+        }else{
+            this.arg=arg;
+        }
+        // isEditArgValid(arg);
+    }
```
Comment (fanyiii) [inline src/seedu/addressbook/commands/EditCommand.java]: Many white spaces are required but missed out here. You may refer to [Java coding standard](https://oss-generic.github.io/process/codingStandards/CodingStandard-Java.html) for more examples (Under Layout - 12).

File: src/seedu/addressbook/commands/EditCommand.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+//        this.toEdit = toEdit;
+//    }
+
+    @Override
+    public CommandResult execute() {
+        //   final List<ReadOnlyPerson> personsFound = getPersonsWithNameContainingAnyKeyword(keywords);
+        // return new CommandResult(getMessageForPersonListShownSummary(personsFound), personsFound);
+        try {
+            final ReadOnlyPerson target = getTargetPerson();
+            getPerson = new Person(target);
```
Comment (fanyiii) [inline src/seedu/addressbook/commands/EditCommand.java]: the variable could be better named to make it less confusing. getPerson sounds more like a method name instead

File: src/seedu/addressbook/commands/FindCommand.java [hunk 6]
```diff
@@ -49,12 +49,21 @@ public CommandResult execute() {
...
+            Set<String> wordsInName = new HashSet<>(person.getName().getWordsInName());
+            wordsInName=lowerCase(wordsInName);
+            keywords=lowerCase(keywords);
             if (!Collections.disjoint(wordsInName, keywords)) {
                 matchedPersons.add(person);
             }
-        }
+    }
         return matchedPersons;
+}
```
Comment (fanyiii) [inline src/seedu/addressbook/commands/FindCommand.java]: Should add 4 spaces to the last two curly brackets

File: src/seedu/addressbook/commands/FindCommand.java [hunk 7]
```diff
@@ -49,12 +49,21 @@ public CommandResult execute() {
...
+    private Set<String> lowerCase(Set<String> list){
+        Set<String> lowerCaseNames = new HashSet<>();
+        for(String names: list){
+            lowerCaseNames.add(names.toLowerCase());
+        }
+        return lowerCaseNames;
     }
```
Comment (fanyiii) [inline src/seedu/addressbook/commands/FindCommand.java]: Good job in implementing your own utility function here. Note that 'names' should be renamed to 'name' instead to represent a single name.

File: src/seedu/addressbook/data/AddressBook.java [hunk 8]
```diff
@@ -86,6 +85,64 @@ public void addPerson(Person toAdd) throws DuplicatePersonException {
     /**
      * Returns true if an equivalent person exists in the address book.
      */
+    public Person editPerson(Person toEdit, String arg, int targetIndex) throws IllegalValueException {
+
+            String[] splitArg = arg.split(" ");
+            final Set<Tag> tagSet = new HashSet<>();
```
Comment (fanyiii) [inline src/seedu/addressbook/data/AddressBook.java]: Indentation problem here?. Our coding standard requires you to use 4 spaces instead of tabs. You can configure IntelliJ to convert tabs to spaces automatically.

File: src/seedu/addressbook/data/AddressBook.java [hunk 9]
```diff
@@ -86,6 +85,64 @@ public void addPerson(Person toAdd) throws DuplicatePersonException {
...
+                        if (i == splitArg.length - 1) {
+                            toEdit.setTags(new UniqueTagList(tagSet));
+                        }
+                        break;
+                }
+
+            }
+        allPersons.edit(toEdit,targetIndex-1);
+
+
```
Comment (fanyiii) [inline src/seedu/addressbook/data/AddressBook.java]: Try to avoid adding unnecessary empty lines

File: src/seedu/addressbook/data/AddressBook.java [hunk 10]
```diff
@@ -86,6 +85,64 @@ public void addPerson(Person toAdd) throws DuplicatePersonException {
     /**
      * Returns true if an equivalent person exists in the address book.
      */
+    public Person editPerson(Person toEdit, String arg, int targetIndex) throws IllegalValueException {
```
Comment (fanyiii) [inline src/seedu/addressbook/data/AddressBook.java]: A better name for this method might be getEditedPerson, since a Person object is returned.

File: src/seedu/addressbook/parser/Parser.java [hunk 11]
```diff
@@ -11,17 +11,8 @@
...
-import seedu.addressbook.commands.Command;
-import seedu.addressbook.commands.DeleteCommand;
-import seedu.addressbook.commands.ExitCommand;
-import seedu.addressbook.commands.FindCommand;
-import seedu.addressbook.commands.HelpCommand;
-import seedu.addressbook.commands.IncorrectCommand;
-import seedu.addressbook.commands.ListCommand;
-import seedu.addressbook.commands.ViewAllCommand;
-import seedu.addressbook.commands.ViewCommand;
+import seedu.addressbook.commands.*;
```
Comment (fanyiii) [inline src/seedu/addressbook/parser/Parser.java]: Again, import statement should be stated explicitly.

File: src/seedu/addressbook/parser/Parser.java [hunk 12]
```diff
@@ -219,25 +243,82 @@ private Command prepareViewAll(String args) {
...
+     * @return the parsed index number
+     * @throws ParseException if no region of the args string could be found for the index
+     */
+//    private static int parseEditArgsAsDisplayedIndex(String args) throws ParseException, NumberFormatException {
+//        final Matcher matcher = PERSON_INDEX_ARGS_FORMAT.matcher(args.trim());
+//        if (!matcher.matches()) {
+//            throw new ParseException("Could not find index number to parse");
+//        }
+//        return Integer.parseInt(matcher.group("targetIndex"));
+//    }
```
Comment (fanyiii) [inline src/seedu/addressbook/parser/Parser.java]: Don't leave commented out code. Just go ahead and delete.

File: src/seedu/addressbook/parser/Parser.java [hunk 13]
```diff
@@ -219,25 +243,82 @@ private Command prepareViewAll(String args) {
...
 
     /**
      * Parses arguments in the context of the find person command.
      *
      * @param args full command args string
      * @return the prepared command
      */
-    private Command prepareFind(String args) {
+    private static Command prepareFind(String args) {
         final Matcher matcher = KEYWORDS_ARGS_FORMAT.matcher(args.trim());
```
Comment (fanyiii) [inline src/seedu/addressbook/parser/Parser.java]: May I know why did you change all the methods to static? A Parser object is created in Main class to parse the command, hence they should be left as instance methods.
